### PR TITLE
Skip checks for overridden or hidden members in overload resolution with extension methods

### PR DESCRIPTION
**Pull request by cston**
Improves Intellisense performance by ~35% for scenario in #67926.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:823]: n=1000: before 48 secs, after 12 secs
n=2500: before 11 mins, after 66 secs

File: src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:862]: n=1000: before 78 secs, after 2 secs
n=2500: before 19 mins, after 9 secs

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:181]: Should we confirm that all containing types are static types derived from `System.Object`?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4), assuming CI is passing

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 7)

**Comment by cston** [comment]
@dotnet/roslyn-compiler for a second review, thanks.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:894]: any value in verifying diagnostics here?

File: src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:956]: Does the language specify this?

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:182]: nit: it feels like this might be easier to understand as `!isExtensionMethodResolution || !members.All(...)`

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:181]: Is this optimization only applicable for extension method overload resolution? What about normal static method overload resolution?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:956]: I'm not sure if the language specifies this, but this is the existing behavior so I wanted to avoid an unnecessary breaking change.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:182]: To me, both forms seem equally difficult to read. I'll leave it as is since the current form matches the comment above which seems relatively clear.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:181]: I limited the optimization to extension methods because it was easy to catch that case in the caller. I think it's less likely there is an issue with static method overloads since the containing types in that case would need to be in the same class or base classes, so few containing types and, as a result, few entries in the dictionary returned from `PartitionMembersByContainingType()`.

**Pull request merged**
